Core/clim-basic/utils.lisp

```diff
@@ -743,3 +728,195 @@ a flag CLOSED is T then beginning and end of the list are consecutive too."
+                   ((= next-opportunity-index (length opportunities))
+                    (return-from breaks-rec
+                      (values (%line-breaks-1 string width offset margin start end) 0)))
+                   ((>= start (aref opportunities next-opportunity-index))
+                    (return-from breaks-rec
+                      (breaks-rec offset margin start end (1+ next-opportunity-index)))))
+             (loop
+                with best-breaks = nil
+                with best-cost = nil
+                for i from next-opportunity-index below (length opportunities)
+                for opportunity = (aref opportunities i)
+                for current-width = (funcall width-fn string start opportunity)
+                do
+                  (cond ((<= current-width current-margin)
+                         (let ((current-cost (expt (- current-margin current-width) 2))
+                               (current-break (if-let ((pos (skip-whitespace string opportunity)))
+                                                pos
+                                                (return-from breaks-rec nil))))
+                           (multiple-value-bind (breaks remaining-cost)
+                               (breaks-rec 0 margin current-break end (1+ i))
+                             (when (or (null best-cost)
+                                       (< (+ current-cost remaining-cost) best-cost))
+                               (setf best-breaks (list* current-break breaks)
+                                     best-cost (+ current-cost remaining-cost))))))
+                        ((null best-cost)
+                         (return-from breaks-rec
+                           (if (> offset 0)
+                               (multiple-value-bind (breaks remaining-cost)
+                                   (breaks-rec 0 margin start end i)
+                                 (return-from breaks-rec
+                                   (values (list* start breaks)
+                                           (+ remaining-cost current-margin))))
+                               (let* ((char-breaks (%line-breaks-1 string width
+                                                                   offset margin
+                                                                   start opportunity))
+                                      (new-start (alexandria:last-elt char-breaks)))
+                                 (multiple-value-bind (breaks remaining-cost)
+                                     (breaks-rec 0 margin new-start end i)
+                                   (values (append char-breaks breaks) remaining-cost))))))
+                        (T #1=(return-from breaks-rec
+                                (values best-breaks best-cost))))
+                finally #1#)))
+    (breaks-rec initial-offset margin start end 0)))
+
+;; Implementing line breaking as defined in Unicode[1] is left as an excercise
```

Comment:
Either remove this comment or explain what the downsides of the algorithm below compared the Unicode algorithm are.

Reply:
addressed in the newest commit